File: doc2markdown/scripts/doc2markdown.py
#!/usr/bin/env python3
"""
doc2markdown
基于docchain远程服务实现多种格式文件转Markdown
默认输出目录为源文件同级目录
"""
import io
import os
import sys
import time
import json
import zipfile

import requests
from typing import Optional

if hasattr(sys.stdout, "reconfigure"):
    sys.stdout.reconfigure(encoding="utf-8", errors="backslashreplace")

POLL_INTERVAL = 3   # 轮询间隔（秒）
POLL_TIMEOUT = 60   # 自动等待上限（秒）


class Doc2Markdown:

    BASE_URL = "https://lab.hjcloud.com/llmdoc/v1"

    def upload_file(self, file_path: str) -> Optional[str]:
        """上传文件到docchain服务，返回文档引用ID"""
        if not os.path.exists(file_path):
            print(f"错误: 文件不存在 - {file_path}")
            return None

        try:
            ascii_filename = os.path.basename(file_path)
            if not ascii_filename:
                ascii_filename = 'document.docx'

            with open(file_path, 'rb') as f:
                files = {'file': (ascii_filename, f)}
                url = f"{self.BASE_URL}/skills/doc2markdown/convert"

                response = requests.post(url, files=files, timeout=30)

                if response.status_code == 200:
                    try:
                        response_json = response.json()
                        if not response_json.get("success", True):
                            print(f"API请求失败: {response_json.get('err')}")
                            return None
                        doc_id = response_json.get('doc_id')
                        if doc_id:
                            return doc_id
                        else:
                            print(f"上传成功但未获取到文档引用ID")
                            return None
                    except json.JSONDecodeError:
                        print(f"响应转换失败，响应内容: {response.text}")
                        return None
                else:
                    print(f"上传失败，状态码: {response.status_code}")
                    print(f"错误信息: {response.text}")
                    return None

        except Exception as e:
            print(f"上传文件时发生错误: {str(e)}")
            return None

    def check_status(self, doc_id: str) -> Optional[str]:
        """
        检查文档处理状态

        Returns:
            'done' | 'failed' | 'converting' | None(请求失败)
        """
        try:
            url = f"{self.BASE_URL}/skills/doc2markdown/check"
            params = {'doc_id': doc_id}

            response = requests.get(url, params=params, timeout=30)

            if response.status_code == 200:
                data = response.json()
                if not data.get("success", True):
                    print(f"检查状态失败: {data.get('err')}")
                    return None
                status_detail = data.get('status_detail', {})
                convert_status = status_detail.get('convert_md', {}).get('state')
                if convert_status == '1':
                    return 'done'
                elif convert_status == '3':
                    return 'failed'
                return 'converting'
            else:
                print(f"检查状态失败，状态码: {response.status_code}，错误信息: {response.text}")
                return None

        except Exception as e:
            print(f"检查状态时发生错误: {str(e)}")
            return None

    def get_markdown(self, doc_id: str):
        """获取转换后的markdown内容（zip包bytes）"""
        try:
            url = f"{self.BASE_URL}/skills/doc2markdown/download"
            params = {'doc_id': doc_id}

            response = requests.get(url, params=params, timeout=30)

            if response.status_code == 200:
                return response.content
            else:
                print(f"获取内容失败，状态码: {response.status_code}")
                return None

        except Exception as e:
            print(f"获取内容时发生错误: {str(e)}")
            return None

    def save_markdown(self, zip_bytes, doc_id: str, file_path: str) -> Optional[str]:
        """解压zip到源文件同级目录，返回输出目录路径"""
        try:
            parent_dir = os.path.dirname(os.path.abspath(file_path))
            file_id, _ = doc_id.split('-')
            markdown_dir_name = f"{file_id}_" + os.path.splitext(os.path.basename(file_path))[0]
            out_dir = os.path.join(parent_dir, markdown_dir_name)
            os.makedirs(out_dir, exist_ok=True)
            safe_out_dir = os.path.realpath(out_dir) + os.sep
            with zipfile.ZipFile(io.BytesIO(zip_bytes)) as z:
                for member in z.infolist():
                    member_path = os.path.realpath(os.path.join(out_dir, member.filename))
                    if not member_path.startswith(safe_out_dir):
                        raise Exception(f"不安全的ZIP条目路径（路径穿越）: {member.filename}")
                    z.extract(member, out_dir)
            return out_dir
        except Exception as e:
            print(f"保存文件时发生错误: {str(e)}")
            return None

    def poll_until_done(self, doc_id: str, file_path: str = None):
        """
        轮询等待转换完成（最多POLL_TIMEOUT秒）

        Returns:
            (True, out_dir)  - 转换完成并已下载
            (False, msg)     - 转换失败
            (None, None)     - 超时未完成
        """
        elapsed = 0
        while elapsed < POLL_TIMEOUT:
            status = self.check_status(doc_id)
            if status is None:
                print("错误: 无法获取文档状态，请稍后重试")
                sys.exit(1)
            if status == 'done':
                print(f"  转换完成，正在下载...")
                zip_bytes = self.get_markdown(doc_id)
                if not zip_bytes:
                    return False, "获取markdown内容失败"
                hint = file_path or f"doc_{doc_id}.md"
                out_dir = self.save_markdown(zip_bytes, doc_id, hint)
                if not out_dir:
                    return False, "保存文件失败"
                return True, out_dir
            elif status == 'failed':
                return False, "文档转换失败"
            else:
                print(f"  转换中... 已等待 {elapsed}s.")
                time.sleep(POLL_INTERVAL)
                elapsed += POLL_INTERVAL

        return None, None

    def convert_file(self, file_path: str):
        """上传文件并自动等待转换，超时则返回doc_id供后续查询"""
        # 1. 上传文件
        print(f"[1/3] 正在上传文件: {file_path}")
        doc_id = self.upload_file(file_path)
        if not doc_id:
            print(f"文件上传失败！")
            sys.exit(1)
        print(f"  上传成功，文档ID: {doc_id}")

        # 2. 轮询等待（最多60秒）
        print(f"[2/3] 等待转换（自动检查）...")
        result, detail = self.poll_until_done(doc_id, file_path)

        if result is True:
            print(f"[3/3] 下载完成，文件已保存: {detail}")
        elif result is False:
            print(f"错误: {detail}")
            sys.exit(1)
        else:
            print(f"  文档ID: {doc_id}")
            print(f"[!] 文档仍在转换中（已超过 {POLL_TIMEOUT}s）")
            print("[!] 大文档通常需要更多时间，如遇系统高峰期需要排队转换，请耐心等待")
            print("[!] 是否需要继续查询转换状态并下载")
            sys.exit(2)

    def check_and_download(self, doc_id: str, file_path: str):
        """通过doc_id检查状态并下载"""
        print(f"[1/2] 检查文档 {doc_id} 的转换状态...")
        status = self.check_status(doc_id)
        if status is None:
            print("错误: 无法获取文档状态，请稍后重试")
            sys.exit(1)

        if status == 'failed':
            print("错误: 文档转换失败")
            sys.exit(1)
        elif status == 'done':
            print(f"  转换已完成，正在下载...")
            zip_bytes = self.get_markdown(doc_id)
            if not zip_bytes:
                print("获取markdown内容失败")
                sys.exit(1)
            out_dir = self.save_markdown(zip_bytes, doc_id, file_path)
            if not out_dir:
                sys.exit(1)
            print(f"[2/2] 下载完成，文件已保存: {out_dir}")
        else:
            print("  文档仍在转换中，继续等待...")
            result, detail = self.poll_until_done(doc_id, file_path)
            if result is True:
                print(f"[2/2] 下载完成，文件已保存: {detail}")
            elif result is False:
                print(f"错误: {detail}")
                sys.exit(1)
            else:
                print(f"[!] 文档仍在转换中，请稍后再试")
                print(f"  文档ID: {doc_id}")
                sys.exit(2)


USAGE = """用法:
  python doc2markdown.py convert <文件路径>          上传并转换文档
  python doc2markdown.py check  <文档ID> <文件路径>   查询状态并下载

示例:
  python doc2markdown.py convert report.pdf
  python doc2markdown.py check 123-f3ce07 report.pdf"""


def main():
    if len(sys.argv) < 2:
        print(USAGE)
        sys.exit(1)

    converter = Doc2Markdown()
    cmd = sys.argv[1]

    if cmd == "convert":
        if len(sys.argv) != 3:
            print("用法: python doc2markdown.py convert <文件路径>")
            sys.exit(1)
        converter.convert_file(sys.argv[2])

    elif cmd == "check":
        if len(sys.argv) != 4:
            print("用法: python doc2markdown.py check <文档ID> <文件路径>")
            sys.exit(1)
        doc_id = sys.argv[2]
        file_path = sys.argv[3]
        converter.check_and_download(doc_id, file_path)

    else:
        # 兼容：直接传文件路径视为 convert
        if os.path.isfile(cmd):
            converter.convert_file(cmd)
        else:
            print(f"未知命令: {cmd}")
            print(USAGE)
            sys.exit(1)


if __name__ == '__main__':
    main()

File: doc2markdown/scripts/doc2markdown.js
#!/usr/bin/env node
/**
 * doc2markdown
 * 基于docchain远程服务实现多种格式文件转Markdown
 * 默认输出目录为源文件同级目录
 */
const fs = require('fs');
const path = require('path');
const http = require('http');
const https = require('https');
const { URL } = require('url');
const { exec } = require('child_process');
const { promisify } = require('util');

const execAsync = promisify(exec);

const POLL_INTERVAL = 3;   // 轮询间隔（秒）
const POLL_TIMEOUT = 60;   // 自动等待上限（秒）

class Doc2Markdown {
    constructor() {
        this.BASE_URL = "http://10.10.185.13:7100/llmdoc/v1";
    }

    /**
     * 发送HTTP请求
     * @param {string} url
     * @param {object} options
     * @returns {Promise<{status: number, data: any}>}
     */
    async request(url, options = {}) {
        return new Promise((resolve, reject) => {
            const urlObj = new URL(url);
            const isHttps = urlObj.protocol === 'https:';
            const httpModule = isHttps ? https : http;

            const reqOptions = {
                hostname: urlObj.hostname,
                port: urlObj.port || (isHttps ? 443 : 80),
                path: urlObj.pathname + urlObj.search,
                method: options.method || 'GET',
                headers: options.headers || {},
                timeout: options.timeout || 30000
            };

            const req = httpModule.request(reqOptions, (res) => {
                const chunks = [];

                res.on('data', (chunk) => {
                    chunks.push(chunk);
                });

                res.on('end', () => {
                    const buffer = Buffer.concat(chunks);
                    let data;

                    if (options.responseType === 'arraybuffer') {
                        data = buffer;
                    } else {
                        const text = buffer.toString('utf8');
                        try {
                            data = JSON.parse(text);
                        } catch (e) {
                            data = text;
                        }
                    }

                    resolve({
                        status: res.statusCode,
                        data: data
                    });
                });
            });

            req.on('error', (error) => {
                reject(error);
            });

            req.on('timeout', () => {
                req.destroy();
                reject(new Error('Request timeout'));
            });

            if (options.body) {
                req.write(options.body);
            }

            req.end();
        });
    }

    /**
     * 生成multipart/form-data格式的请求体
     * @param {string} filePath
     * @param {string} filename
     * @returns {{body: Buffer, boundary: string}}
     */
    createMultipartFormData(filePath, filename) {
        const boundary = '----WebKitFormBoundary' + Math.random().toString(36).substring(2);
        const fileContent = fs.readFileSync(filePath);

        const parts = [];

        // 添加文件字段
        parts.push(
            `--${boundary}\r\n` +
            `Content-Disposition: form-data; name="file"; filename="${filename}"\r\n` +
            `Content-Type: application/octet-stream\r\n\r\n`
        );

        parts.push(fileContent);
        parts.push(`\r\n--${boundary}--\r\n`);

        // 合并所有部分
        const buffers = parts.map(part =>
            Buffer.isBuffer(part) ? part : Buffer.from(part, 'utf8')
        );

        const body = Buffer.concat(buffers);

        return { body, boundary };
    }

    /**
     * 上传文件到docchain服务，返回文档引用ID
     * @param {string} filePath
     * @returns {Promise<string|null>}
     */
    async uploadFile(filePath) {
        if (!fs.existsSync(filePath)) {
            console.log(`错误: 文件不存在 - ${filePath}`);
            return null;
        }

        try {
            let asciiFilename = path.basename(filePath);
            if (!asciiFilename) {
                asciiFilename = 'document.docx';
            }

            const { body, boundary } = this.createMultipartFormData(filePath, asciiFilename);
            const url = `${this.BASE_URL}/skills/doc2markdown/convert`;

            const response = await this.request(url, {
                method: 'POST',
                headers: {
                    'Content-Type': `multipart/form-data; boundary=${boundary}`,
                    'Content-Length': body.length
                },
                body: body,
                timeout: 30000
            });

            if (response.status === 200) {
                const responseJson = response.data;
                if (!responseJson.success && responseJson.success !== undefined) {
                    console.log(`API请求失败: ${responseJson.err}`);
                    return null;
                }
                const docId = responseJson.doc_id;
                if (docId) {
                    return docId;
                } else {
                    console.log(`上传成功但未获取到文档引用ID`);
                    return null;
                }
            } else {
                console.log(`上传失败，状态码: ${response.status}`);
                console.log(`错误信息: ${JSON.stringify(response.data)}`);
                return null;
            }

        } catch (error) {
            console.log(`上传文件时发生错误: ${error.message}`);
            return null;
        }
    }

    /**
     * 检查文档处理状态
     * @param {string} docId
     * @returns {Promise<'done'|'failed'|'converting'|null>}
     */
    async checkStatus(docId) {
        try {
            const url = `${this.BASE_URL}/skills/doc2markdown/check?doc_id=${encodeURIComponent(docId)}`;

            const response = await this.request(url, {
                method: 'GET',
                timeout: 30000
            });

            if (response.status === 200) {
                const data = response.data;
                if (!data.success && data.success !== undefined) {
                    console.log(`检查状态失败: ${data.err}`);
                    return null;
                }
                const statusDetail = data.status_detail || {};
                const convertStatus = (statusDetail.convert_md || {}).state;
                if (convertStatus === '1') {
                    return 'done';
                } else if (convertStatus === '3') {
                    return 'failed';
                }
                return 'converting';
            } else {
                console.log(`检查状态失败，状态码: ${response.status}，错误信息: ${JSON.stringify(response.data)}`);
                return null;
            }

        } catch (error) {
            console.log(`检查状态时发生错误: ${error.message}`);
            return null;
        }
    }

    /**
     * 获取转换后的markdown内容（zip包bytes）
     * @param {string} docId
     * @returns {Promise<Buffer|null>}
     */
    async getMarkdown(docId) {
        try {
            const url = `${this.BASE_URL}/skills/doc2markdown/download?doc_id=${encodeURIComponent(docId)}`;

            const response = await this.request(url, {
                method: 'GET',
                timeout: 30000,
                responseType: 'arraybuffer'
            });

            if (response.status === 200) {
                return response.data;
            } else {
                console.log(`获取内容失败，状态码: ${response.status}`);
                return null;
            }

        } catch (error) {
            console.log(`获取内容时发生错误: ${error.message}`);
            return null;
        }
    }

    /**
     * 解压zip到源文件同级目录，返回输出目录路径
     * @param {Buffer} zipBytes
     * @param {string} docId
     * @param {string} filePath
     * @returns {Promise<string|null>}
     */
    async saveMarkdown(zipBytes, docId, filePath) {
        try {
            const parentDir = path.dirname(path.resolve(filePath));
            const [fileId] = docId.split('-');
            const markdownDirName = `${fileId}_` + path.parse(filePath).name;
            const outDir = path.join(parentDir, markdownDirName);

            if (!fs.existsSync(outDir)) {
                fs.mkdirSync(outDir, { recursive: true });
            }

            const zlib = require('zlib');
            const { pipeline } = require('stream/promises');

            // 临时保存 zip
            const tempZip = path.join(outDir, `temp_${Date.now()}.zip`);
            fs.writeFileSync(tempZip, zipBytes);

            // 原生解析 ZIP 中央目录
            const buffer = fs.readFileSync(tempZip);
            let pos = buffer.length - 22;
            while (pos > 0) {
                if (
                    buffer.readUInt32LE(pos) === 0x06054b50 &&
                    pos + 22 <= buffer.length
                )
                    break;
                pos--;
            }

            const entries = [];
            const diskEntries = buffer.readUInt16LE(pos + 8);
            const dirStart = buffer.readUInt32LE(pos + 16);
            pos = dirStart;

            for (let i = 0; i < diskEntries; i++) {
                if (buffer.readUInt32LE(pos) !== 0x02014b50) break;
                const flags = buffer.readUInt16LE(pos + 8);
                const method = buffer.readUInt16LE(pos + 10);
                const nameLen = buffer.readUInt16LE(pos + 28);
                const extraLen = buffer.readUInt16LE(pos + 30);
                const commentLen = buffer.readUInt16LE(pos + 32);
                const offset = buffer.readUInt32LE(pos + 42);
                const name = buffer.toString('utf8', pos + 46, pos + 46 + nameLen);
                entries.push({ offset, method, name, encrypted: !!(flags & 1) });
                pos += 46 + nameLen + extraLen + commentLen;
            }

            // 解压每个文件
            for (const ent of entries) {
                if (ent.encrypted || ent.name.endsWith('/')) continue;
                const o = ent.offset;
                const sig = buffer.readUInt32LE(o);
                if (sig !== 0x04034b50) continue;
                const nameLen = buffer.readUInt16LE(o + 26);
                const extraLen = buffer.readUInt16LE(o + 28);
                const csize = buffer.readUInt32LE(o + 18);
                const usize = buffer.readUInt32LE(o + 14);
                const dataStart = o + 30 + nameLen + extraLen;
                const data = buffer.slice(dataStart, dataStart + csize);

                const outPath = path.join(outDir, ent.name.replace(/\\/g, '/'));
                const outDirPath = path.dirname(outPath);
                if (!fs.existsSync(outDirPath)) fs.mkdirSync(outDirPath, { recursive: true });

                if (ent.method === 0) {
                    fs.writeFileSync(outPath, data);
                } else if (ent.method === 8) {
                    const decompressed = zlib.inflateSync(data, { chunkSize: usize });
                    fs.writeFileSync(outPath, decompressed);
                }
            }

            fs.unlinkSync(tempZip);
            return outDir;

        } catch (error) {
            console.log(`保存文件时发生错误: ${error.message}`);
            return null;
        }
    }

    /**
     * 轮询等待转换完成（最多POLL_TIMEOUT秒）
     * @param {string} docId
     * @param {string|null} filePath
     * @returns {Promise<[boolean, string]|[null, null]>}
     */
    async pollUntilDone(docId, filePath = null) {
        let elapsed = 0;
        while (elapsed < POLL_TIMEOUT) {
            const status = await this.checkStatus(docId);
            if (status === null) {
                console.log("错误: 无法获取文档状态，请稍后重试");
                process.exit(1);
            }
            if (status === 'done') {
                console.log(`  转换完成，正在下载...`);
                const zipBytes = await this.getMarkdown(docId);
                if (!zipBytes) {
                    return [false, "获取markdown内容失败"];
                }
                const hint = filePath || `doc_${docId}.md`;
                const outDir = await this.saveMarkdown(zipBytes, docId, hint);
                if (!outDir) {
                    return [false, "保存文件失败"];
                }
                return [true, outDir];
            } else if (status === 'failed') {
                return [false, "文档转换失败"];
            } else {
                console.log(`  转换中... 已等待 ${elapsed}s.`);
                await new Promise(resolve => setTimeout(resolve, POLL_INTERVAL * 1000));
                elapsed += POLL_INTERVAL;
            }
        }

        return [null, null];
    }

    /**
     * 上传文件并自动等待转换，超时则返回doc_id供后续查询
     * @param {string} filePath
     */
    async convertFile(filePath) {
        // 1. 上传文件
        console.log(`[1/3] 正在上传文件: ${filePath}`);
        const docId = await this.uploadFile(filePath);
        if (!docId) {
            console.log(`文件上传失败！`);
            process.exit(1);
        }
        console.log(`  上传成功，文档ID: ${docId}`);

        // 2. 轮询等待（最多60秒）
        console.log(`[2/3] 等待转换（自动检查）...`);
        const [result, detail] = await this.pollUntilDone(docId, filePath);

        if (result === true) {
            console.log(`[3/3] 下载完成，文件已保存: ${detail}`);
        } else if (result === false) {
            console.log(`错误: ${detail}`);
            process.exit(1);
        } else {
            console.log(`  文档ID: ${docId}`);
            console.log(`[!] 文档仍在转换中（已超过 ${POLL_TIMEOUT}s）`);
            console.log(`[!] 大文档通常需要更多时间，如遇系统高峰期需要排队转换，请耐心等待`);
            console.log(`[!] 是否需要继续查询转换状态并下载`);
            process.exit(2);
        }
    }

    /**
     * 通过doc_id检查状态并下载
     * @param {string} docId
     * @param {string} filePath
     */
    async checkAndDownload(docId, filePath) {
        console.log(`[1/2] 检查文档 ${docId} 的转换状态...`);
        const status = await this.checkStatus(docId);
        if (status === null) {
            console.log("错误: 无法获取文档状态，请稍后重试");
            process.exit(1);
        }

        if (status === 'failed') {
            console.log("错误: 文档转换失败");
            process.exit(1);
        } else if (status === 'done') {
            console.log(`  转换已完成，正在下载...`);
            const zipBytes = await this.getMarkdown(docId);
            if (!zipBytes) {
                console.log("获取markdown内容失败");
                process.exit(1);
            }
            const outDir = await this.saveMarkdown(zipBytes, docId, filePath);
            if (!outDir) {
                process.exit(1);
            }
            console.log(`[2/2] 下载完成，文件已保存: ${outDir}`);
        } else {
            console.log("  文档仍在转换中，继续等待...");
            const [result, detail] = await this.pollUntilDone(docId, filePath);
            if (result === true) {
                console.log(`[2/2] 下载完成，文件已保存: ${detail}`);
            } else if (result === false) {
                console.log(`错误: ${detail}`);
                process.exit(1);
            } else {
                console.log(`[!] 文档仍在转换中，请稍后再试`);
                console.log(`  文档ID: ${docId}`);
                process.exit(2);
            }
        }
    }
}

const USAGE = `用法:
  node doc2markdown-native.js convert <文件路径>          上传并转换文档
  node doc2markdown-native.js check  <文档ID> <文件路径>   查询状态并下载

示例:
  node doc2markdown-native.js convert report.pdf
  node doc2markdown-native.js check 123-f3ce07 report.pdf`;


async function main() {
    if (process.argv.length < 3) {
        console.log(USAGE);
        process.exit(1);
    }

    const converter = new Doc2Markdown();
    const cmd = process.argv[2];

    if (cmd === "convert") {
        if (process.argv.length !== 4) {
            console.log("用法: node doc2markdown-native.js convert <文件路径>");
            process.exit(1);
        }
        await converter.convertFile(process.argv[3]);

    } else if (cmd === "check") {
        if (process.argv.length !== 5) {
            console.log("用法: node doc2markdown-native.js check <文档ID> <文件路径>");
            process.exit(1);
        }
        const docId = process.argv[3];
        const filePath = process.argv[4];
        await converter.checkAndDownload(docId, filePath);

    } else {
        // 兼容：直接传文件路径视为 convert
        if (fs.existsSync(cmd) && fs.statSync(cmd).isFile()) {
            await converter.convertFile(cmd);
        } else {
            console.log(`未知命令: ${cmd}`);
            console.log(USAGE);
            process.exit(1);
        }
    }
}

if (require.main === module) {
    main().catch(error => {
        console.error(`程序执行错误: ${error.message}`);
        process.exit(1);
    });
}

module.exports = Doc2Markdown;